lecture-notes/2-vcs.md

```diff
@@ -1574,9 +1574,251 @@ were easy to edit commits, `git revert` would likely work just the same.
+  fprintf(stdout, "%d\n", result);
+  return 0;
+}
+$ 
+```
+
+This retrieves the same past version of `main.c` from the Git history
+but---instead of printing it---writes it directly to the working tree. Unlike
+`git revert`, it does not create a new commit or stage the changes, meaning they
+show up in `git status` and `git diff` like if you'd made them manually.
+
+There's also a third way to go back in history---`git checkout`. Like `git
+restore`, `git checkout` changes the working tree. But it doesn't leave those
+changes unstaged; instead, it moves `HEAD` to point to the given commit,
+resulting in no unstaged *or* staged changes---since Git shows changes relative
+to `HEAD`!
+
+```console?prompt=$
+$ git checkout 2d6603026105
+Note: switching to '2d6603026105'.
+
+You are in 'detached HEAD' state. You can look around, make experimental
+changes and commit them, and you can discard any commits you make in this
+state without impacting any branches by switching back to a branch.
+
+If you want to create a new branch to retain commits you create, you may
+do so (now or later) by using -c with the switch command. Example:
+
+  git switch -c <new-branch-name>
+
+Or undo this operation with:
+
+  git switch -
+
+Turn off this advice by setting config variable advice.detachedHead to false
+
+HEAD is now at 2d66030 Support the modulo operator
+$ cat main.c
+#include <errno.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+int calc(int left, char op, int right) {
+  switch (op) {
+  case '+':
+    return left + right;
+  case '-':
+    return left - right;
+  case '*':
+    return left * right;
+  case '/':
+    return left / right;
+  case  '%':
+    return left % right;
+  }
+  fprintf(stderr, "Unrecognized op `%c'.\n", op);
+  exit(EXIT_FAILURE);
+}
+
+int main(int argc, char **argv) {
+  if (argc != 4) {
+    fprintf(
+        stderr,
+        "Usage: %s <num> <op> <num>\nWhere <op> is one of +, -, *, /, %%.\n",
+        argv[0]);
+    return EXIT_FAILURE;
+  }
+  const char *left_str = argv[1];
+  const char *op_str = argv[2];
+  const char *right_str = argv[3];
+  errno = 0;
+  int left = strtol(left_str, NULL, 10);
+  if (errno != 0) {
+    perror(argv[0]);
+    return EXIT_FAILURE;
+  }
+  if (op_str[1] != '\0') {
+    fprintf(stderr, "Op must be one character. Got `%s'.\n", op_str);
+    return EXIT_FAILURE;
+  }
+  errno = 0;
+  int right = strtol(right_str, NULL, 10);
+  if (errno != 0) {
+    perror(argv[0]);
+    return EXIT_FAILURE;
+  }
+  int result = calc(left, op_str[0], right);
+  fprintf(stdout, "%d\n", result);
+  return 0;
+}
+$ git status
+HEAD detached at 2d66030
+nothing to commit, working tree clean
+$ 
+```
+
+As Git helpfully informs us, checking out a raw commit hash isn't the typical
+use of `git checkout`, as it it puts your repository into a state called
+"detached HEAD", where new commits aren't tracked as part of a branch. Until we
+discuss branches next lecture, don't worry too much about that: just run `git
```

Comment:
I think we can make a tiiiiny bit more noise about "this is the big error that scares people"